File: backend/app/schemas/problem.py
"""
题目相关的Pydantic模式
"""
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field, validator
from datetime import datetime

# 题目基础模式
class ProblemBase(BaseModel):
    """题目基础模式"""
    title: str = Field(..., min_length=1, max_length=200, description="题目标题")
    content: str = Field(..., description="题目内容")
    content_type: str = Field(default="text", description="内容类型: text, markdown, latex")
    options: Dict[str, str] = Field(..., description="题目选项，如{'A': '选项A', 'B': '选项B'}")
    correct_answer: str = Field(..., pattern='^[A-D]$', description="正确答案: A, B, C, D")
    solution: Optional[str] = None
    solution_type: str = Field(default="text", description="解析类型: text, markdown")
    difficulty: int = Field(default=3, ge=1, le=5, description="难度等级: 1-5")
    source_type: Optional[str] = None
    source_year: Optional[int] = None
    source_detail: Optional[str] = None
    estimated_time: Optional[int] = Field(None, ge=1, description="预估时间（秒）")
    
    @validator('options')
    def validate_options(cls, v):
        """验证选项格式"""
        required_keys = {'A', 'B', 'C', 'D'}
        if not required_keys.issubset(v.keys()):
            missing = required_keys - set(v.keys())
            raise ValueError(f"缺少选项: {', '.join(missing)}")
        return v

# 创建题目
class ProblemCreate(ProblemBase):
    """创建题目模式"""
    knowledge_point_ids: List[int] = Field(default=[], description="关联的知识点ID列表")
    is_published: bool = Field(default=False, description="是否立即发布")

# 更新题目
class ProblemUpdate(BaseModel):
    """更新题目模式"""
    title: Optional[str] = Field(None, min_length=1, max_length=200)
    content: Optional[str] = None
    content_type: Optional[str] = None
    options: Optional[Dict[str, str]] = None
    correct_answer: Optional[str] = Field(None, pattern='^[A-D]$')
    solution: Optional[str] = None
    solution_type: Optional[str] = None
    difficulty: Optional[int] = Field(None, ge=1, le=5)
    source_type: Optional[str] = None
    source_year: Optional[int] = None
    source_detail: Optional[str] = None
    estimated_time: Optional[int] = Field(None, ge=1)
    is_published: Optional[bool] = None
    knowledge_point_ids: Optional[List[int]] = None
    
    @validator('options')
    def validate_options(cls, v):
        if v is not None:
            required_keys = {'A', 'B', 'C', 'D'}
            if not required_keys.issubset(v.keys()):
                missing = required_keys - set(v.keys())
                raise ValueError(f"缺少选项: {', '.join(missing)}")
        return v

# 题目响应
class ProblemResponse(ProblemBase):
    """题目响应模式"""
    id: int
    is_published: bool
    review_status: str
    total_attempts: int
    correct_attempts: int
    accuracy_rate: float
    created_by: int
    created_at: datetime
    updated_at: Optional[datetime] = None
    
    class Config:
        from_attributes = True

# 题目详情
class ProblemDetail(ProblemResponse):
    """题目详情模式"""
    knowledge_points: Optional[List[Dict[str, Any]]] = None
    creator: Optional[Dict[str, Any]] = None

# 题目列表查询
class ProblemFilter(BaseModel):
    """题目过滤条件"""
    difficulty: Optional[List[int]] = None
    source_type: Optional[str] = None
    source_year: Optional[int] = None
    knowledge_point_id: Optional[int] = None
    is_published: Optional[bool] = True
    search: Optional[str] = None
    sort_by: str = "created_at"
    sort_order: str = "desc"

# 题目统计
class ProblemStats(BaseModel):
    """题目统计模式"""
    total_problems: int
    published_problems: int
    by_difficulty: Dict[int, int]
    by_source: Dict[str, int]
    avg_accuracy: float

# 练习题目（不包含答案）
class PracticeProblem(BaseModel):
    """练习用题目模式（隐藏答案）"""
    id: int
    title: str
    content: str
    content_type: str
    options: Dict[str, str]
    difficulty: int
    estimated_time: Optional[int] = None
    knowledge_points: Optional[List[Dict[str, Any]]] = None

File: frontend/src/types/problem.ts
/**
 * 题目内容类型
 */
export type ContentType = 'text' | 'markdown' | 'latex';

/**
 * 题目来源类型
 */
export type SourceType = 'AMC8' | '迎春杯' | '华杯赛' | string; // 允许其他自定义来源

/**
 * 审核状态
 */
export type ReviewStatus = 'pending' | 'approved' | 'rejected';


/**
 * 题目选项接口（JSONB 格式）
 */
export interface ProblemOptions {
  [key: string]: string; // 例如: { "A": "选项A的内容", "B": "选项B的内容" }
}

/**
 * 题目统计信息
 */
export interface ProblemStatistics {
  total_attempts: number;
  correct_attempts: number;
  success_rate?: number; // 计算字段，可根据前两个字段计算
}

/**
 * 题目基础接口
 */
export interface ProblemBase {
  // 基本信息
  id: number;
  title: string;
  content: string;
  content_type: ContentType;
  
  // 选项
  options: ProblemOptions;
  
  // 答案和解析
  correct_answer: string; // 'A', 'B', 'C', 'D'
  solution?: string;
  solution_type?: ContentType;
  
  // 难度和分类
  difficulty: number; // 1-5
  source_type?: SourceType;
  source_year?: number;
  source_detail?: string;
  
  // 预估属性
  estimated_time?: number; // 秒
  success_rate?: number; // 0-1 的小数
  
  // 状态控制
  is_published: boolean;
  is_deleted: boolean;
  
  // 审核信息
  reviewed_by?: number;
  reviewed_at?: Date;
  review_status: ReviewStatus;
  
  // 统计信息
  total_attempts: number;
  correct_attempts: number;
  
  // 创建信息
  created_by?: number;
  created_at: Date;
  updated_at: Date;
  
  // 全文搜索字段（PostgreSQL 特有）
  search_vector?: any;
}

/**
 * 题目完整接口（包含关联的用户信息）
 */
export interface Problem extends ProblemBase {
  // 关联的用户信息（可选，根据需要扩展）
  creator_name?: string;
  reviewer_name?: string;
}

/**
 * 创建题目请求接口
 */
export interface CreateProblemRequest {
  title: string;
  content: string;
  content_type?: ContentType;
  options: ProblemOptions;
  correct_answer: string;
  solution?: string;
  solution_type?: ContentType;
  difficulty: number;
  source_type?: SourceType;
  source_year?: number;
  source_detail?: string;
  estimated_time?: number;
  is_published?: boolean;
}

/**
 * 更新题目请求接口
 */
export interface UpdateProblemRequest extends Partial<CreateProblemRequest> {
  review_status?: ReviewStatus;
  is_deleted?: boolean;
}

/**
 * 题目查询参数接口
 */
export interface ProblemQueryParams {
  // 分页参数
  page?: number;
  limit?: number;
  
  // 筛选参数
  difficulty?: number | number[];
  source_type?: SourceType | SourceType[];
  source_year?: number | number[];
  is_published?: boolean;
  review_status?: ReviewStatus | ReviewStatus[];
  created_by?: number;
  
  // 搜索参数
  search?: string; // 全文搜索关键词
  title?: string; // 标题模糊搜索
  
  // 排序参数
  sort_by?: 'id' | 'title' | 'difficulty' | 'created_at' | 'updated_at' | 'success_rate';
  sort_order?: 'asc' | 'desc';
}

/**
 * 题目列表响应接口
 */
export interface ProblemListResponse {
  problems: Problem[];
  total: number;
  page: number;
  limit: number;
  total_pages: number;
}

/**
 * 题目统计响应接口
 */
export interface ProblemStats {
  total: number;
  published: number;
  pending_review: number;
  by_difficulty: {
    [difficulty: number]: number;
  };
  by_source_type: {
    [sourceType: string]: number;
  };
}

/**
 * 题目难度枚举（前端显示用）
 */
export const DIFFICULTY_LABELS: Record<number, string> = {
  1: '非常简单',
  2: '简单',
  3: '中等',
  4: '困难',
  5: '非常困难'
};

/**
 * 审核状态枚举（前端显示用）
 */
export const REVIEW_STATUS_LABELS: Record<ReviewStatus, string> = {
  pending: '待审核',
  approved: '已通过',
  rejected: '已拒绝'
};

/**
 * 题目来源类型枚举（前端显示用）
 */
export const SOURCE_TYPE_LABELS: Record<string, string> = {
  'AMC8': '美国数学竞赛8',
  '迎春杯': '迎春杯数学竞赛',
  '华杯赛': '华罗庚金杯数学竞赛'
};

/**
 * 计算题目的正确率
 */
export function calculateSuccessRate(
  correctAttempts: number, 
  totalAttempts: number
): number {
  if (totalAttempts === 0) return 0;
  return Math.round((correctAttempts / totalAttempts) * 10000) / 10000; // 保留4位小数
}

/**
 * 获取难度文本
 */
export function getDifficultyText(difficulty: number): string {
  return DIFFICULTY_LABELS[difficulty] || `难度 ${difficulty}`;
}

/**
 * 获取审核状态文本
 */
export function getReviewStatusText(status: ReviewStatus): string {
  return REVIEW_STATUS_LABELS[status] || status;
}

/**
 * 获取来源类型文本
 */
export function getSourceTypeText(sourceType?: string): string {
  if (!sourceType) return '未知来源';
  return SOURCE_TYPE_LABELS[sourceType] || sourceType;
}

/**
 * 题目过滤器类型
 */
export interface ProblemFilter {
  difficulties?: number[];
  sourceTypes?: string[];
  years?: number[];
  searchText?: string;
  onlyPublished?: boolean;
}

/**
 * 题目答案提交接口
 */
export interface SubmitAnswerRequest {
  problemId: number;
  answer: string;
  timeSpent?: number; // 秒
}

/**
 * 题目答案响应接口
 */
export interface SubmitAnswerResponse {
  is_correct: boolean;
  correct_answer: string;
  solution?: string;
  solution_type?: ContentType;
  explanation?: string;
  statistics?: {
    current_success_rate: number;
    total_attempts: number;
    correct_attempts: number;
  };
}

/**
 * 题目详情扩展接口（包含用户尝试记录）
 */
export interface ProblemWithAttempt extends Problem {
  user_attempt?: {
    attempted: boolean;
    is_correct?: boolean;
    user_answer?: string;
    attempted_at?: Date;
    time_spent?: number;
  };
}

/**
 * 题目导入/导出格式
 */
export interface ProblemExportFormat {
  version: string;
  exported_at: string;
  problems: Array<{
    title: string;
    content: string;
    content_type: ContentType;
    options: ProblemOptions;
    correct_answer: string;
    solution?: string;
    solution_type?: ContentType;
    difficulty: number;
    source_type?: string;
    source_year?: number;
    source_detail?: string;
    estimated_time?: number;
  }>;
}